File: neo/animation/inveq.py
from sympy import *

def psi_c0(alpha, beta, zeta, gamma):
    psi_res = I*log((-sqrt(cos(alpha)**2*cos(beta)**2 -\
    2*cos(alpha)*cos(beta)*cos(gamma)*cos(zeta) +\
    cos(gamma)**2 + cos(zeta)**2 - 1) +\
    cos(alpha)*cos(beta)*cos(zeta) -\
    cos(gamma))/((sin(alpha)*cos(beta) -\
    I*sin(beta))*sin(zeta)))
    evalv_psi = psi_res.evalf()
    print("PSI = ",end=''); pprint(evalv_psi)
    return float(re(evalv_psi))

def phi(alpha, beta, zeta, gamma):
    before_asin = (cos(alpha)*cos(beta) -\
        cos(gamma)*cos(zeta))/(sin(gamma)*sin(zeta))
    ic(before_asin)
    phi_res = asin((cos(alpha)*cos(beta) -\
        cos(gamma)*cos(zeta))/(sin(gamma)*sin(zeta)))
    evalv_phi = phi_res.evalf()
    print("PHI = ",end=''); pprint(evalv_phi)
    return float(evalv_phi)

File: neo/animation/inveq.m


function phi = phi_eq(alpha,beta,lambda,gamma)
    % Inverse kinematic equation for Y-Joint
    phi = asin( (cos(alpha) * cos(beta) - cos(gamma)*cos(lambda)) / ...
        (sin(gamma)*sin(lambda) ) )
end

function psi = psi_eq(alpha,beta,lambda,gamma)
    % Complex inverse kinematic equation for X-Joint
    psi = 1j*log((-sqrt(cos(alpha)^2*cos(beta)^2 -\
        2*cos(alpha)*cos(beta)*cos(gamma)*cos(lambda) +\
        cos(gamma)^2 + cos(lambda)^2 - 1) +\
        cos(alpha)*cos(beta)*cos(lambda) -\
        cos(gamma))/((sin(alpha)*cos(beta) -\
        1j*sin(beta))*sin(lambda)))
end

function psi_vec = psi_vec_eq(alpha,beta,lambda,gamma)
    % Vector analysis inverse kinematic equation for X-Joint
    psi_vec = asin((-(sin(gamma)*sin(phi)*cos(lambda) -...
        sin(lambda)*cos(gamma))*sin(beta) +...
        sin(alpha)*sin(gamma)*cos(beta)*cos(phi))/...
        (sqrt(-cos(alpha)^2*cos(beta)^2 + 1)*...
        sqrt((sin(gamma)*sin(phi)*cos(lambda) -...
        sin(lambda)*cos(gamma))^2 +...
        sin(gamma)^2*cos(phi)^2)))
